cpp/src/arrow/sparse_tensor.h

```diff
@@ -29,42 +29,58 @@ namespace arrow {
 // ----------------------------------------------------------------------
 // SparseIndex class
 
-/// \brief EXPERIMENTAL: Sparse tensor format enumeration
 struct SparseTensorFormat {
+  /// EXPERIMENTAL: The index format type of SparseTensor
   enum type { COO, CSR };
 };
 
-/// \brief EXPERIMENTAL: The base class for representing index of non-zero
-/// values in sparse tensor
+/// \brief EXPERIMENTAL: The base class for the index of a sparse tensor
+///
+/// SparseIndex are used to represent where is non-zero values in a
+/// SparseTensor.
+///
+/// There are several ways of the index representation.  The format_id is used
+/// to distinguish what kind of representation is used.  And the concrete
+/// implementation will be done in subclasses.
 class ARROW_EXPORT SparseIndex {
  public:
   explicit SparseIndex(SparseTensorFormat::type format_id, int64_t non_zero_length)
       : format_id_(format_id), non_zero_length_(non_zero_length) {}
 
   virtual ~SparseIndex() = default;
 
+  /// \brief Return the identifier of the format type
   SparseTensorFormat::type format_id() const { return format_id_; }
+
+  /// \brief Return the number of non zero values in the sparse tensor related
+  /// to this sparse index
   int64_t non_zero_length() const { return non_zero_length_; }
 
+  /// \brief Return the string representation of the sparse index
   virtual std::string ToString() const = 0;
 
  protected:
   SparseTensorFormat::type format_id_;
   int64_t non_zero_length_;
 };
 
-template <typename SparseIndexType>
-class SparseIndexBase : public SparseIndex {
- public:
-  explicit SparseIndexBase(int64_t non_zero_length)
-      : SparseIndex(SparseIndexType::format_id, non_zero_length) {}
-};
+namespace internal {
+  template <typename SparseIndexType>
+  class SparseIndexBase : public SparseIndex {
+   public:
+    explicit SparseIndexBase(int64_t non_zero_length)
+        : SparseIndex(SparseIndexType::format_id, non_zero_length) {}
+  };
+}
```

Comment:
I put this class in `internal` namespace to remove it from documentation targets.